### DISCUSSION TITLE
Setting up Python interface for openEMS and CSXCAD

### DISCUSSION THREAD

**Discussion by adityamalhotraAU** [thliebig/openEMS-Project#167] [Q&A Usage]
Hello, I am very new to openEMS and I have been having troubles setting up the python interface for CSXCAD and openEMS. I followed the instructions given on the website for setting this up. However when I am trying to use pip to install the required python libraries it is throwing a /usr/local/clang exited with code 1 for the CSXCAD python interface. It also says,

/Users/adityamalhotra/Library/Caches/Homebrew/openems--git/CSXCAD/python/CSXCAD/CSPrimitives.cpp:1222:10: fatal error: 'CSXCAD/ParameterObjects.h' file not found
 #include "CSXCAD/ParameterObjects.h"

I checked and the ParameterObjects.h file is not present inside the CSXCAD directory anywhere. It also says "python setup.py bdist_wheel did not run successfully".

<img width="1470" alt="Screenshot 2023-12-17 at 8 35 32 AM" src="https://github.com/thliebig/openEMS-Project/assets/150011222/ba5fe4fe-3aed-488e-b27b-109096ca9471">

When trying to setup the python interface for openEMS, the error thrown is "Error compiling cython file". It also says "python setup.py egg_info did not run successfully". For this the file that is being compiled is trying to cimport cython.numeric. However it is throwing the error cython.numeric is not a valid cython.* module.

<img width="1470" alt="Screenshot 2023-12-17 at 8 36 08 AM" src="https://github.com/thliebig/openEMS-Project/assets/150011222/213f4d9f-8885-45ef-8d89-35b6a74fdf72">

I have already installed the required python libraries but it has not worked.

**Comment by biergaizi** [comment]
First of all, please *never* post text as screenshot when you're reporting an issue for *any* software project. Screenshots cannot be edited or searched like text, so they're wastes of everyone's time. For short error messages like these, it only takes a few seconds to copy the entire console output as text and press the "code" formatting button. For long error messages, use attachments.

> However it is throwing the error cython.numeric is not a valid cython.* module.

This has just been fixed several days ago. The patch can be applied manually from here: https://github.com/thliebig/openEMS/commit/bf30b7d94ab63bbe63672dd0651172410df4b044 - but it's *not* enough to solve your problem because there's also a CSXCAD problem.

> error: 'CSXCAD/ParameterObjects.h' file not found

I'm the author of the documentation, but I don't understand why this happens, It worked on my machine. What macOS version are you using?

Anyway, it sounds like a problem related to compiler search path. For now, try copying all `.h` files from `CSXCAD/src` to `CSXCAD/python/CSXCAD/`. If it's still not working, try copying them to `CSXCAD/python/CSXCAD/CSXCAD/` (you need to create it).

CC @aWZHY0yQH81uOYvH: Please take a look when you have time.

**Reply by aWZHY0yQH81uOYvH** [reply]
I am seeing this problem too. The C++ files generated during the cythonizing process have include paths like `CSXCAD/ParameterObjects.h`. Nowhere in the CSXCAD repository is that header file in a parent directory called `CSXCAD`, so I assume that it is intending to use the header files that you have installed on your system. On an Intel Mac using Homebrew, that path is `/usr/local/include/CSXCAD/ParameterObjects.h`. However, when built using Homebrew-installed python:

```
/usr/local/bin/python3 setup.py install --user
```

(again I'm on an Intel Mac so my brew prefix is `/usr/local`)

The resulting compilation command is as follows:

```
clang -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -I/usr/local/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11 -c /tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSPrimitives.cpp -o build/temp.macosx-14-x86_64-cpython-311/tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSPrimitives.o -v
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: x86_64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple x86_64-apple-macosx14.0.0 -Wundef-prefix=TARGET_OS_ -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -emit-obj --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name CSPrimitives.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=all -mllvm -extra-vectorizer-passes -fno-strict-return -ffp-contract=on -fno-rounding-math -funwind-tables=2 -target-sdk-version=14.2 -fvisibility-inlines-hidden-static-local-var -target-cpu penryn -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=standalone -dwarf-version=4 -debugger-tuning=lldb -target-linker-version 1022.1 -v -fcoverage-compilation-dir=/tmp/openEMS-Project/CSXCAD/python -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -D NDEBUG -I /usr/local/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11 -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include/c++/v1 -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/local/include -internal-isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include -internal-externc-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include -internal-externc-isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include -O3 -Wsign-compare -Wunreachable-code -Wall -Wno-reorder-init-list -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-final-dtor-non-final-class -Wno-extra-semi-stmt -Wno-misleading-indentation -Wno-quoted-include-in-framework-header -Wno-implicit-fallthrough -Wno-enum-enum-conversion -Wno-enum-float-conversion -Wno-elaborated-enum-base -Wno-reserved-identifier -Wno-gnu-folding-constant -fdeprecated-macro -fdebug-compilation-dir=/tmp/openEMS-Project/CSXCAD/python -ferror-limit 19 -fwrapv -stack-protector 1 -fstack-check -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fno-cxx-modules -fcxx-exceptions -fexceptions -fmax-type-align=16 -fcolor-diagnostics -vectorize-loops -vectorize-slp -clang-vendor-feature=+disableNonDependentMemberExprInCurrentInstantiation -fno-odr-hash-protocols -clang-vendor-feature=+enableAggressiveVLAFolding -clang-vendor-feature=+revert09abecef7bbf -clang-vendor-feature=+thisNoAlignAttr -clang-vendor-feature=+thisNoNullAttr -mllvm -disable-aligned-alloc-awareness=1 -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o build/temp.macosx-14-x86_64-cpython-311/tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSPrimitives.o -x c++ /tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSPrimitives.cpp
clang -cc1 version 15.0.0 (clang-1500.1.0.2.5) default target x86_64-apple-darwin23.2.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include/c++/v1
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks (framework directory)
End of search list.
/tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSPrimitives.cpp:1225:10: fatal error: 'CSXCAD/ParameterObjects.h' file not found
#include "CSXCAD/ParameterObjects.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
error: command '/usr/bin/clang' failed with exit code 1
```

The include search paths do not include `/usr/local/include`, so clang can't find the installed CSXCAD headers. It appears that cython uses whatever compiler flags it was built with, and I believe the main culprit here is `-isysroot` which overwrites the normal system root paths.

If you attempt to build using the system-provided python, it works.

```
/usr/bin/python3 setup.py install --user
```

```
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration -Wno-error=unreachable-code -I/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -c /tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSProperties.cpp -o build/temp.macosx-10.9-universal2-3.9/tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSProperties.o -v
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: x86_64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple arm64-apple-macosx10.9.0 -Wundef-prefix=TARGET_OS_ -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -emit-obj --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name CSProperties.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=non-leaf -mllvm -extra-vectorizer-passes -fno-strict-return -ffp-contract=on -fno-rounding-math -funwind-tables=1 -faligned-alloc-unavailable -fobjc-msgsend-selector-stubs -target-sdk-version=14.2 -fvisibility-inlines-hidden-static-local-var -target-cpu apple-m1 -target-feature +v8.5a -target-feature +crc -target-feature +lse -target-feature +rdm -target-feature +crypto -target-feature +dotprod -target-feature +fp-armv8 -target-feature +neon -target-feature +fp16fml -target-feature +ras -target-feature +rcpc -target-feature +zcm -target-feature +zcz -target-feature +fullfp16 -target-feature +sm4 -target-feature +sha3 -target-feature +sha2 -target-feature +aes -target-abi darwinpcs -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=standalone -dwarf-version=4 -debugger-tuning=lldb -target-linker-version 1022.1 -v -fcoverage-compilation-dir=/tmp/openEMS-Project/CSXCAD/python -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0 -iwithsysroot /System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -D NDEBUG -I /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -I/usr/local/include -internal-isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1 -internal-isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include -internal-isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include -internal-externc-isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -internal-externc-isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include -O3 -Wno-unused-result -Wsign-compare -Wunreachable-code -Wall -Werror=implicit-function-declaration -Wno-error=unreachable-code -Wno-reorder-init-list -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-final-dtor-non-final-class -Wno-extra-semi-stmt -Wno-misleading-indentation -Wno-quoted-include-in-framework-header -Wno-implicit-fallthrough -Wno-enum-enum-conversion -Wno-enum-float-conversion -Wno-elaborated-enum-base -Wno-reserved-identifier -Wno-gnu-folding-constant -fdeprecated-macro -fdebug-compilation-dir=/tmp/openEMS-Project/CSXCAD/python -ferror-limit 19 -fwrapv -stack-protector 1 -fstack-check -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fno-cxx-modules -fcxx-exceptions -fexceptions -fmax-type-align=16 -fcolor-diagnostics -vectorize-loops -vectorize-slp -clang-vendor-feature=+disableNonDependentMemberExprInCurrentInstantiation -fno-odr-hash-protocols -clang-vendor-feature=+enableAggressiveVLAFolding -clang-vendor-feature=+revert09abecef7bbf -clang-vendor-feature=+thisNoAlignAttr -clang-vendor-feature=+thisNoNullAttr -mllvm -disable-aligned-alloc-awareness=1 -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /var/folders/m2/vn4yhpk53fg8c28hm8422p1c0000gn/T/CSProperties-029269/CSProperties-arm64.o -x c++ /tmp/openEMS-Project/CSXCAD/python/CSXCAD/CSProperties.cpp
clang -cc1 version 15.0.0 (clang-1500.1.0.2.5) default target x86_64-apple-darwin23.2.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/System.framework/PrivateHeaders"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
```

Notice the use of `-iwithsysroot`, which adds a system root instead of overwriting it.

To fix this, either you need to build with the system-supplied version of python (by explicitly specifying `/usr/bin/python3` or equivalent on the command line or by changing around your `PATH`), or by adding your system include paths to `setup.py` as follows. @biergaizi I assume you may have the non-default (but security-wise advantageous) configuration of `/usr/bin` being before `/usr/local/bin` in your `PATH`, so that you are using a non-brew-installed version of python by default.

```diff
diff --git a/python/setup.py b/python/setup.py
index c8c6c05..994360e 100644
--- a/python/setup.py
+++ b/python/setup.py
@@ -5,6 +5,7 @@ import os
 
 extensions = [
     Extension("*", [os.path.join(os.path.dirname(__file__), "CSXCAD","*.pyx")],
+        include_dirs = ["/usr/local/include"],
         language="c++",             # generate C++ code
         libraries = ['CSXCAD',]),
 ]
```

A better permanent solution to this would be to figure out a cross-platform way of finding the default system include paths and specifying them as above, modifying the flags provided to clang (perhaps by replacing `-isysroot` with `-iwithsysroot`; see [here](https://stackoverflow.com/questions/57046796/how-to-remove-pthread-compiler-flag-from-cython-setup-file/57057959#57057959)), or by modifying the directory structure in `CSXCAD/python/CSXCAD` to have a symlink or something that allows finding the header files naturally.

**Reply by aWZHY0yQH81uOYvH** [reply]
I have opened issues regarding this. The problem is present in openEMS as well.
https://github.com/thliebig/CSXCAD/issues/38
https://github.com/thliebig/openEMS/issues/135

**Reply by biergaizi** [reply]
Excellent troubleshooting, thank you.

> To fix this, either you need to build with the system-supplied version of python (by explicitly specifying /usr/bin/python3 or equivalent on the command line or by changing around your PATH), or by adding your system include paths to setup.py as follows. @biergaizi I assume you may have the non-default (but security-wise advantageous) configuration of /usr/bin being before /usr/local/bin in your PATH, so that you are using a non-brew-installed version of python by default.

Yes. I just remembered seeing the same problem as well with Homebrew's Python, but I thought that it was a non-default environment anyway so supporting it was not strictly necessary, so I simply uninstalled Homebrew's Python in favor of macOS's built-in one and didn't continue investigating. If I recall correctly, if openEMS is the only thing you've installed, Homebrew won't pull its own Python as a dependency, so everything worked. But naturally this is not a typical setup and we really need to solve the real underlying problem.

**Comment by nicoceledonc** [comment]
> fatal error: 'CSXCAD/ParameterObjects.h' file not found

Got the same `fatal error: 'CSXCAD/ParameterObjects.h' file not found`but not sure how to approach. I tried

- adding `include_dirs = ["/usr/local/include"],`
- running with `/usr/bin/python3 setup.py install --user`

I think that it's because I have conda in the middle but not sure

Have a mac M1 and Sequoia 15.1

**Reply by snhobbs** [reply]
Is CSXCAD in your /usr/local/include/? Otherwise you'll need to add the path your headers are located or copy them there. 

Also I highly recommend a docker image for this, the dependencies were a big issue for me.

**Reply by nicoceledonc** [reply]
Thanks for the quick reply, but I finally use a macport install. Not an experienced user at all on openEMS so haven't been able to check if it works correctly, but it seems to be okay.

**Reply by snhobbs** [reply]
Can you leave a short description of what you did in case future Mac users stumble over here?

**Reply by SengerM** [reply]
Just for reference, not long ago [someone was having similar issues](https://github.com/thliebig/openEMS-Project/issues/241#issuecomment-2453578044) and [the PR with the `pip` installation method](https://github.com/thliebig/openEMS-Project/pull/239) seems to have solved it. Would be nice to have more tests.
